src/solvers/dg_p4est/containers_3d.jl

```diff
@@ -0,0 +1,303 @@
+          contravariant_vectors, inverse_jacobian = elements
+
+  calc_node_coordinates!(node_coordinates, mesh, basis.nodes)
+
+  for element in 1:ncells(mesh)
+    calc_jacobian_matrix!(jacobian_matrix, element, node_coordinates, basis)
+
+    calc_contravariant_vectors!(contravariant_vectors, element, jacobian_matrix,
+                                node_coordinates, basis)
+
+    calc_inverse_jacobian!(inverse_jacobian, element, jacobian_matrix, basis)
+  end
+
+  return nothing
+end
+
+
+# Interpolate tree_node_coordinates to each quadrant
+function calc_node_coordinates!(node_coordinates,
+                                mesh::P4estMesh{3},
+                                basis::LobattoLegendreBasis)
+  # Hanging nodes will cause holes in the mesh if its polydeg is higher
+  # than the polydeg of the solver.
+  @assert length(basis.nodes) >= length(mesh.nodes) "The solver can't have a lower polydeg than the mesh"
+
+  calc_node_coordinates!(node_coordinates, mesh, basis.nodes)
+end
+
+function calc_node_coordinates!(node_coordinates,
+                                mesh::P4estMesh{3},
+                                nodes::AbstractVector)
+  # Macros from p4est
+  p4est_root_len = 1 << P4EST_MAXLEVEL
+  p4est_quadrant_len(l) = 1 << (P4EST_MAXLEVEL - l)
+
+  trees = unsafe_wrap_sc(p8est_tree_t, mesh.p4est.trees)
+
+  for tree in eachindex(trees)
+    offset = trees[tree].quadrants_offset
+    quadrants = unsafe_wrap_sc(p8est_quadrant_t, trees[tree].quadrants)
+
+    for i in eachindex(quadrants)
+      element = offset + i
+      quad = quadrants[i]
+
+      quad_length = p4est_quadrant_len(quad.level) / p4est_root_len
+
+      nodes_out_x = 2 * (quad_length * 1/2 * (nodes .+ 1) .+ quad.x / p4est_root_len) .- 1
+      nodes_out_y = 2 * (quad_length * 1/2 * (nodes .+ 1) .+ quad.y / p4est_root_len) .- 1
+      nodes_out_z = 2 * (quad_length * 1/2 * (nodes .+ 1) .+ quad.z / p4est_root_len) .- 1
+
+      matrix1 = polynomial_interpolation_matrix(mesh.nodes, nodes_out_x)
+      matrix2 = polynomial_interpolation_matrix(mesh.nodes, nodes_out_y)
+      matrix3 = polynomial_interpolation_matrix(mesh.nodes, nodes_out_z)
+
+      multiply_dimensionwise!(
+        view(node_coordinates, :, :, :, :, element),
+        matrix1, matrix2, matrix3,
+        view(mesh.tree_node_coordinates, :, :, :, :, tree)
+      )
+    end
+  end
+
+  return node_coordinates
+end
+
+
+@inline function init_interface_node_indices!(interfaces::InterfaceContainerP4est{3},
+                                              faces, orientation, interface_id)
+  # Iterate over primary and secondary element
+  for side in 1:2
+    # Align interface at the primary element (primary element has surface indices (:i, :j)).
+    # The secondary element needs to be indexed differently.
+    if side == 1
+      surface_index1 = :i
+      surface_index2 = :j
+    else
+      surface_index1, surface_index2 = p4est_orientation_to_indices(faces[2], faces[1], orientation)
```

Comment:
Shouldn't this function be called `orientation_to_indices_p4est`? I think it would be good to avoid creating functions prefixed with `p4est_`, since this complicates the situation for users who will not know if it is a p4est library function or a Trixi function until they look it up.